add option to make response buffers temporary

bokwoon95 opened the pull request:
Thanks for this plugin, it's really useful to me. However, whenever I have a malformed request I don't wish to retain the response buffer, and they really pollute the buffer list!

I added an option `g:vim_http_tempbuffer` that makes all response buffers temporary so they leave no trace once their window is closed. Whenever you send a new `:Http` request, all open response windows get closed first before a new response window is opened. This makes it easy to keep tweaking the request until you get the response you want, at which you can then copy the response to another file or something to preserve the output.

[![asciicast](https://asciinema.org/a/OtJaOUovIoTMEoWirk3ww7hbs.svg)](https://asciinema.org/a/OtJaOUovIoTMEoWirk3ww7hbs)

This can be made into a toggleable On/Off thing, but I'm not sure how you'd prefer it to be implemented.

bokwoon95 committed:
add option to make response buffers temporary

When g:vim_http_tempbuffer is set, response buffers created are
temporary and do not get added to the buffer list.

bokwoon95 committed:
Preserve alternate buffer when g:vim_http_tempbuffer is active

When switching to the temporary response buffer, do not mark it as an
alternate buffer. This is so that once the response window is closed
(and hence buffer deleted), pressing <C-6> or :b# will not trigger E23:
No alternate file

nicwest commented on autoload/http.vim:
```diff
@@ -151,12 +151,26 @@ endfunction
+    "   execute 'vert new ' . l:buffer_name
+    " else
+    "   execute 'new ' . l:buffer_name
+    " end
```
You can go ahead and remove this stuff, we have version control if we ever need it back.

nicwest commented on autoload/http.vim:
```diff
@@ -151,12 +151,26 @@ endfunction
-    end
+    if g:vim_http_tempbuffer
+      for win in range(1, winnr('$'))
+        if getwinvar(win, 'temp')
```
is `w:temp` a builtin or a standard thing to do? I can't see anything obvious in help.

If not can you prefix the name to `w:http_vim_temp` or similar to ensure that we don't clobber some buffer also setting `w:temp` from another plugin.

nicwest reviewed (commented):
LGTM, couple of minor questions/changes inline.

nicwest commented:
also can you stick a test in somewhere for this?

bokwoon95 committed:
Fix PR issues

1) Give window variable more specific name to prevent name conflicts
2) Remove commented out code

bokwoon95 committed:
add test for g:vim_http_tempbuffer

bokwoon95 commented:
I'm not very familiar with adding tests yet, could you check if it seems on the right track?

bokwoon95 commented:
Oh man I'm not sure why 
```
E481: No range allowed: 1windo close
```
is happening, `:windo` does take a range when I try it manually in vim

nicwest commented:
Let me take a look, it could be a CI thing, the version of vim running there is a bit weird.

nicwest commented:
```
Patch 7.4.566
Problem:    :argdo, :bufdo, :windo and :tabdo don't take a range.
Solution:   Support the range. (Marcin Szamotulski)
Files:      runtime/doc/editing.txt, runtime/doc/tabpage.txt,
            runtime/doc/windows.txt, src/ex_cmds.h, src/ex_cmds2.c,
            src/testdir/test_command_count.in,
            src/testdir/test_command_count.ok
```
https://vimhelp.org/version8.txt.html#version8.txt

```
$ vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Nov 24 2016 16:42:52)
Included patches: 1-52
Extra patches: 8.0.0056
```

https://travis-ci.org/nicwest/vim-http/builds/505534838#L574

Yeah looks like that feature was added after the version on travis (ubuntu trusty).

If there is an easy way to do this without breaking support for old versions of vim then lets do that, if not then I'll look into upgrading travis to a more recent version.

nicwest commented on autoload/http.vim:
```diff
@@ -151,12 +151,21 @@ endfunction
+    if g:vim_http_tempbuffer
+      for win in range(1, winnr('$'))
+        if getwinvar(win, 'vim_http_tempbuffer')
+          execute win . 'windo close'
```
`l:win` here please, I don't know how this escaped the linter.

bokwoon95 committed:
Changed `win` variable scope to function local

bokwoon95 commented:
Odd, I was thinking it would be simple to workaround by jumping into the window and closing it (which should put you back in the previous window). 

<img width="334" alt="image" src="https://user-images.githubusercontent.com/23723357/54330366-5cd06500-4650-11e9-80da-0ec630fadae1.png">

But I'm getting this unexpected behaviour where both windows become the response buffer. I can't see how that would happen at all

[![asciicast](https://asciinema.org/a/9txuDjbwWxUsKEdgxR5KJgO5z.svg)](https://asciinema.org/a/9txuDjbwWxUsKEdgxR5KJgO5z)

bokwoon95 commented:
Instead of breaking support for older versions of vim, how about force-setting `g:vim_http_tempbuffer` to `0` for the versions of vim without the patch? 

I feel like it shouldn't be so difficult to replace `:{range}windo close` with `:wincmd c`, but for some reason it's throwing weird behaviour

bokwoon95 commented:
> Instead of breaking support for older versions of vim, how about force-setting `g:vim_http_tempbuffer` to `0` for the versions of vim without the patch?

Actually this might not be a good idea in case someone decides to manually set it again inside vim.

nicwest commented:
this might work (I suspect the test to still fail but at least we won't be getting an exception):
```diff
diff --git a/autoload/http.vim b/autoload/http.vim
index 713a48e..fc3c25a 100644
--- a/autoload/http.vim
+++ b/autoload/http.vim
@@ -151,10 +151,10 @@ endfunction
 function! s:new_response_buffer(request_buffer, response) abort
     let l:request_buffer_name  = bufname(a:request_buffer)
     let l:buffer_name = fnamemodify(l:request_buffer_name, ":r") . '.response.' . localtime() . '.http'
-    if g:vim_http_tempbuffer
-          execute win . 'windo close'
+    if g:vim_http_tempbuffer && (v:version > 704 || (v:version == 704 && has('patch566')))
       for l:win in range(1, winnr('$'))
         if getwinvar(l:win, 'vim_http_tempbuffer')
+          execute l:win . 'windo close'
         endif
       endfor
     endif
diff --git a/test/integration.vim b/test/integration.vim
index b9679a1..a18b5d4 100644
--- a/test/integration.vim
+++ b/test/integration.vim
@@ -248,7 +248,6 @@ function! s:suite.tempbuffer()
     call s:load_request_expected('simple_get')
     Http
     wincmd p
-    call s:load_request_expected('get_url_params')
     Http
     wincmd c
```

P.s. I had to remove the additional file opening from the test in order for it to pass on my machine.

nicwest commented:
I'm probably going to sit down this weekend and move the tests over to a new version of vim so this problem will also go away then.

bokwoon95 committed:
fixed basic error

nicwest commented:
@bokwoon95 if you rebase from master all the tests should be running against a version of vim that supports the `windo` ranges.

bokwoon95 committed:
Fixed test to handle two http requests

The previous version was prematurely getting the expected buffer list
(at the very start of the function) before opening the two .http
buffers, naturally causing the the assert to mismatch.

Instead, open the two .http buffers first before getting the expected
buffer list, then carrying out the `:Http` tests before comparing it
with the final buffer list.

bokwoon95 commented:
> @bokwoon95 if you rebase from master all the tests should be running against a version of vim that supports the `windo` ranges.

> P.s. I had to remove the additional file opening from the test in order for it to pass on my machine.

Yup, there was something wrong with my test logic so I fixed it in commit 
781206a, now it opens both files and the tests pass.

<img width="561" alt="image" src="https://user-images.githubusercontent.com/23723357/54678041-38d4be00-4b3f-11e9-8f6d-e59e7e023695.png">

nicwest commented:
thanks for this, sorry it was such a pain!
